nova/conductor/manager.py

```diff
@@ -186,6 +198,280 @@ def object_backport_versions(self, context, objinst, object_versions):
+                    # do some cleanup, so another compute-node is used
+                    found_hv_sizes_per_az[rp['az']].remove(hv_size)
+
+        hv_sizes_per_az = {
+            az: set(rp['hv_size'] for rp in vmware_providers.values()
+                    if rp['az'] == az)
+            for az in availability_zones}
+
+        missing_hv_sizes_per_az = {
+            az: hv_sizes_per_az[az] - found_hv_sizes_per_az[az]
+            for az in availability_zones}
+
+        if not any(missing_hv_sizes_per_az.values()):
+            LOG.info('Free host for spawning available for every '
+                     'availability-zone and hypervisor-size.')
+            return
+
+        def _flatten(list_of_lists):
+            return itertools.chain.from_iterable(list_of_lists)
+
+        # retrieve allocation candidates for all hv sizes. we later have to
+        # filter them by AZ, because our placement doesn't know about AZs.
+        candidates = {}
+        for hv_size in _flatten(missing_hv_sizes_per_az.values()):
+            resources = {MEMORY_MB: hv_size}
+            res = client.get_allocation_candidates(context, resources)
+            if res is None:
+                continue
+            alloc_reqs, provider_summaries, allocation_request_version = res
+
+            # filter out providers, that don't match the full host, e.g. don't
+            # allow 3 TB on a 6 TB host, as we need a fully free host
+            provider_summaries = {p: d for p, d in provider_summaries.items()
+                    if vmware_providers.get(p, {}).get('hv_size') == hv_size}
+
+            if not provider_summaries:
+                LOG.warning('Could not find enough resources to free up a '
+                            'host for hypervisor size %(hv_size)d.',
+                            {'hv_size': hv_size})
+                continue
+
+            candidates[hv_size] = (alloc_reqs, provider_summaries)
+
+        for az in availability_zones:
+            for hv_size in missing_hv_sizes_per_az[az]:
+                alloc_reqs, provider_summaries = candidates[hv_size]
+
+                # filter providers by AZ, as placement returned all matching
+                # providers
+                providers = {p: d for p, d in provider_summaries.items()
+                        if vmware_providers.get(p, {}).get('az') == az}
+
+                # select the one with the least usage
+                def _free_memory(p):
+                    memory = providers[p]['resources'][MEMORY_MB]
+                    return memory['capacity'] - memory['used']
+
+                provider_uuids = sorted((p for p in providers),
+                                        key=_free_memory, reverse=True)
+
+                for rp_uuid in provider_uuids:
+                    host = vmware_providers[rp_uuid]['host']
+                    self._free_host_for_provider(context, client, rp_uuid,
+                                                 special_spawn_rpc, host)
```

Comment:
Please correct me if I'm wrong: 
- the providers are taken by AZ and sorted in the descending order of free memory amount
- but then we try to free a host per each provider?
- why does it have to be sorted in such case? - maybe a comment will be helpful to understand

Reply:
You're right. I seem to have forgotten to break out of the loop once I got a successful result from one resource-provider.